### PR TITLE
BREAKING!: feat @cordis/util

### LINKED ISSUE #29: @cordis/util
## Feature Request
A helper library that provides abstraction over writing endpoints for common routes. This will be structured in a FP manner, accepting arguments that will be concatenated into the endpoints. It will also handle having typings for each endpoint body (discord-api-types??).

## Function List:

(Moved to #32)

### LINKED ISSUE #46: rest/router POST should be able to take querystring params
https://discord.com/developers/docs/resources/webhook#execute-webhook
specifically the wait query string param

No clue if this should expand to other http methods

### PR DESCRIPTION
**Pull request by zaida04**
closes #29 
badabing badaboom

edit:
This PR will also house fixes for several rest/router bugs.
closes #46

edit 2 (from @didinele):
This PR will also lead to the doom of `@cordis/snowflake` and breaking changes in `@cordis/common`

### Audit Log

- [x] fetchAuditLog

### Channel

- [x] fetchChannel
- [x] editChannel - encapsulate permissions
- [x] deleteChannel
- [x] fetchChannelMessages/fetchChannelMessage
- [x] sendMessage
- [x] crosspostMessage
- [x] addMessageReaction
- [x] deleteOwnMessageReaction
- [x] deleteUserReaction
- [x] deleteAllReactions
- [x] deleteReaction
- [x] fetchMessageReactions
- [x] editMessage
- [x] deleteMessage
- [x] bulkDeleteMessages
- [x] fetchChannelInvites
- [x] createChannelInvite
- [x] followNewsChannel
- [x] startTyping
- [x] fetchPinnedChannelMessages
- [x] deletePinnedChannelMessage
- [x] addPinnedChannelMessage
- [x] groupDMAddRecipient
- [x] groupDMRemoveRecipient

### Emoji

- [x] fetchGuildEmojis
- [x] fetchGuildEmoji
- [x] createGuildEmoji
- [x] editGuildEmoji
- [x] deleteGuildEmoji

### Guild
- [x] createGuild
- [x] fetchGuild
- [x] fetchGuildPreview
- [x] editGuild
- [x] deleteGuild
- [x] fetchGuildChannels
- [x] createGuildChannel
- [x] fetchGuildMember
- [x] addGuildMember
- [x] editGuildMember
- [x] setNickname
- [x] addRole
- [x] removeRole
- [x] kickMember
- [x] fetchGuildBans
- [x] fetchGuildBan
- [x] banMember
- [x] unbanMember
- [x] fetchRoles
- [x] createRole
- [x] setRolePosition
- [x] editRole
- [x] deleteRole
- [x] pruneGuild
- [x] getPredictedPruneResult
- [x] fetchGuildVoiceRegions
- [x] fetchGuildInvites
- [x] createGuildIntegration
- [x] editGuildIntegration
- [x] deleteGuildIntegration
- [x] syncGuildIntegration
- [x] fetchGuildWidget
- [x] editGuildWidget (settings too)
- [x] fetchGuildVanityURL
- [x] fetchGuildWidgetImage

### Invite
- [x] fetchInvite
- [x] deleteInvite

### Template
- [x] fetchTemplate
- [x] createGuildFromTemplate
- [x] fetchGuildTemplates
- [x] createGuildTemplate
- [x] syncGuildTemplate
- [x] modifyGuildTemplate
- [x] deleteGuildTemplate

### User
- [x] fetchCurrentUser
- [x] fetchUser
- [x] editCurrentUser
- [x] leaveGuild
- [x] fetchDMs
- [x] createDM
- [x] createGroupDM
- [x] getUserConnections

### Webhooks
- [x] createWebhook
- [x] fetchChannelWebhooks
- [x] fetchGuildWebhooks
- [x] fetchWebhook (with token too)
- [x] editWebhook (with token too)
- [x] deleteWebhook (with token too)
- [x] executeWebhook
- [x] editWebhookMessage
- [x] deleteWebhookMessage

### CODE REVIEW

File: libs/util/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    "type": "git",
+    "url": "git+https://github.com/cordis-lib/cordis.git"
+  },
+  "contributors": [
+    "didinele <didinele.dev@gmail.com> (https://github.com/didinele)",
+    "Holo-Buckshot (https://github.com/Holo-Buckshot)",
+    "Nico (https://github.com/zaida04)"
+  ],
+  "author": "Zaid \"Nico\"",
+  "license": "MIT",
```
Comment (didinele) [inline libs/util/package.json:28]: Uh, code base is licensed under Apache 2. Should exclude this field anyway, pointless to have it everywhere.

File: libs/util/package.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@cordis/util",
+  "version": "0.1.0",
```
Comment (didinele) [inline libs/util/package.json:3]: fun fact: mono repos aren't versioned this way. even if no changes were made to a package, it should all be lined up.

File: libs/util/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    "!**/*.test.*"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/cordis-lib/cordis.git"
+  },
+  "contributors": [
+    "didinele <didinele.dev@gmail.com> (https://github.com/didinele)",
+    "Holo-Buckshot (https://github.com/Holo-Buckshot)",
+    "Nico (https://github.com/zaida04)"
```
Comment (didinele) [inline libs/util/package.json:25]: Feel free to add this to every other package as well, I was gonna do it but forgot.

**Review by didinele — CHANGES_REQUESTED** [review]
Oh and also, `pnpm i` to update the lock file so CI passes.

File: libs/util/package.json (hunk 3 above)
Comment (zaida04) [reply libs/util/package.json:25]: I don't know whether I should do it in this pr or not, but to be safe i'll just do it in a separate PR

Comment (zaida04) [reply libs/util/package.json:25]: resolved in https://github.com/cordis-lib/cordis/pull/33/commits/0ed8dcee7774b640a1c7743e2ed91de04ab095a7

File: libs/util/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import type { RestManager } from '@cordis/rest';
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+export function makeRestUtils(restRouter: RestManager) {
```
Comment (didinele) [inline libs/util/src/index.ts:4]:
```suggestion
export const makeRestUtils = (rest: RestManager) => {
```

Prefer arrow functions.

File: libs/util/package.json [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  "devDependencies": {
+    "@types/node": "^14.14.31",
+    "typescript": "^4.2.2"
+  },
+  "dependencies": {
+    "@cordis/rest": "workspace:^0.1.5",
+    "tslib": "^2.1.0"
+  },
+  "homepage": "https://github.com/cordis-lib/cordis#readme"
```
Comment (didinele) [inline libs/util/package.json:39]: This is giving me anxiety, should be after `bugs` and before `devDependencies`

**Review by didinele — CHANGES_REQUESTED** [review]

File: libs/util/src/makeDiscordCdnUrl.test.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import { makeDiscordCdnUrl } from './';
+
+describe('make discord cdn url', () => {
```
Comment (didinele) [inline libs/util/src/makeDiscordCdnUrl.test.ts:3]: Should get rid of that describe layer. It was in the other place to clarify which function its coming from.

**Review by didinele — CHANGES_REQUESTED** [review]

**Comment by didinele** [comment]
Was just writing some code and figured we could also throw a few `BitField` implementations into this package.

Example: https://github.com/cordis-lib/cordis/blob/7cdefd8cba6343a5293107b42bf76632b378699d/libs/gateway/src/Intents.ts

That one it makes sense to keep in `@cordis/gateway`, as to not create a reundant dependency on `@cordis/util`, but otherwise for things like Permissions we could dump 'em here.

If you feel the inconsistency is too much we could move `Intents` to `@cordis/common` and have `@cordis/util` re-export, I wouldn't really mind that dependency chain.

**Comment by zaida04** [comment]
I think keeping Intents in gateway is fine, since it's a gateway only thing. I'm also fine with having BitField implementations like permissions in @cordis/util.

**Comment by didinele** [comment]
Also, also, an embed builder? 👀

**Review by didinele — APPROVED** [review]
So far LGTM

**Review by didinele — CHANGES_REQUESTED** [review]

**Review by didinele — APPROVED** [review]

**Comment by didinele** [comment]
Could you remove `routers` from the ROOT readme? Just noticed that's a problem.

**Review by didinele — DISMISSED** [review]
This is now in a merge-able state 🎉

**Review by didinele — APPROVED** [review]
Re-approve since commit that makes CI pass.

Which reminds me, if you got get the test coverage for embed up before we merge this (Sunday), that'd be awesome.

If you wanna see what's being covered/isn't: run the tests (tip: you can use `pnpm run test "--" --watchAll`) and open coverage/lcov-report/index.html and navigate to the embed file.

**Review by didinele — APPROVED** [review]

**Pull request merged**